mailbox/postgres/src/main/java/org/apache/james/mailbox/jpa/mail/dao/PostgresMailboxDAOImpl.java

```diff
@@ -0,0 +1,176 @@
+import static org.apache.james.mailbox.jpa.mail.table.PostgresMailboxTable.USER_NAME;
+import static org.jooq.impl.DSL.count;
+
+import java.util.NoSuchElementException;
+
+import javax.inject.Inject;
+
+import org.apache.james.backends.postgres.utils.PostgresExecutor;
+import org.apache.james.core.Username;
+import org.apache.james.mailbox.exception.MailboxExistsException;
+import org.apache.james.mailbox.jpa.mail.table.PostgresMailboxTable;
+import org.apache.james.mailbox.jpa.user.PostgresMailboxId;
+import org.apache.james.mailbox.model.Mailbox;
+import org.apache.james.mailbox.model.MailboxId;
+import org.apache.james.mailbox.model.MailboxPath;
+import org.apache.james.mailbox.model.UidValidity;
+import org.apache.james.mailbox.model.search.MailboxQuery;
+import org.jooq.Record1;
+
+import com.google.common.base.Preconditions;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class PostgresMailboxDAOImpl implements PostgresMailboxDAO {
+    private static final char SQL_WILDCARD_CHAR = '%';
+
+    private PostgresExecutor postgresExecutor;
+
+    @Inject
+    public PostgresMailboxDAOImpl(PostgresExecutor postgresExecutor) {
+        this.postgresExecutor = postgresExecutor;
+    }
+
+    @Override
+    public Mono<Mailbox> create(MailboxPath mailboxPath, UidValidity uidValidity) {
+        return postgresExecutor.dslContext()
+            .flatMapMany(dslContext -> Mono.from(dslContext.insertInto(TABLE_NAME, MAILBOX_NAME, USER_NAME, MAILBOX_NAMESPACE, MAILBOX_UID_VALIDITY)
+                .values(mailboxPath.getName(), mailboxPath.getUser().asString(), mailboxPath.getNamespace(), uidValidity.asLong())
+                .returningResult(MAILBOX_ID))
+                .map(record -> new Mailbox(mailboxPath, uidValidity, PostgresMailboxId.of(record.get(MAILBOX_ID)))))
+            .last()
+            .onErrorResume(throwable -> {
+                if (throwable.getMessage().contains("duplicate key value violates unique constraint")) {
+                    return Mono.error(new MailboxExistsException(mailboxPath.getName()));
+                } else {
+                    return Mono.error(throwable);
+                }
+            });
+    }
+
+    @Override
+    public Mono<MailboxId> rename(Mailbox mailbox) {
+        Preconditions.checkNotNull(mailbox.getMailboxId(), "A mailbox we want to rename should have a defined mailboxId");
+
+        return postgresExecutor.dslContext()
+            .flatMap(dslContext -> Mono.from(dslContext.update(TABLE_NAME)
+                .set(MAILBOX_NAME, mailbox.getName())
+                .set(USER_NAME, mailbox.getUser().asString())
+                .set(MAILBOX_NAMESPACE, mailbox.getNamespace())))
+            .thenReturn(mailbox.getMailboxId())
+            .onErrorResume(throwable -> {
+                if (throwable.getMessage().contains("duplicate key value violates unique constraint")) {
```

Comment:
Could extract it into a method, to avoid duplicate error management (I see same block in create method)